### A Look at the Data

In order to get a better understanding of the data we will be looking at throughout this lesson, let's take a look at some of the characteristics of the dataset.

First, let's read in the data and necessary libraries.

In [248]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [249]:
b_cal = pd.read_csv('boston_calendar.csv')
b_list = pd.read_csv('boston_listings.csv')
b_rev = pd.read_csv('boston_reviews.csv')

s_cal = pd.read_csv('seatle_calendar.csv')
s_list = pd.read_csv('seatle_listings.csv')
s_rev = pd.read_csv('seatle_reviews.csv')

## Task 1: Busines Understanding

### Step 1: Basic Exploration with minimal cleaning
*To familiarize with the Data and to gather insights to formulate questions*

> **Boston & Seatle Calendar**

#### Check the sizes of cols and rows 

In [250]:
print(b_cal.head(2)),print('\n'), print(s_cal.head(2))
print('\n')
print("b_cal  :", b_cal.shape)
print("s_cal  :", s_cal.shape)

   listing_id        date available price
0    12147973  2017-09-05         f   NaN
1    12147973  2017-09-04         f   NaN


   listing_id        date available   price
0      241032  2016-01-04         t  $85.00
1      241032  2016-01-05         t  $85.00


b_cal  : (1308890, 4)
s_cal  : (1393570, 4)


#### Check Nulls

In [251]:
print('Boston Cal: ')
print("Cols with nulls: ", b_cal.isnull().sum()[b_cal.isnull().sum()>0].index[0])
print("Null prop of price column: ", b_cal.price.isnull().sum()/b_cal.shape[0])
print("Proportion of False(unit unavailable at this date ):", b_cal.available[b_cal.available =='f' ].count()/b_cal.shape[0])
#Since the proportions are the same, then, all nans are when the units aren't available, let's be more sure
print("Nulls when units are available: ", b_cal[b_cal['available']== 't']['price'].isnull().sum())
print('\n')
print('Seatle Cal: ')
print("Cols with nulls: ", s_cal.isnull().sum()[s_cal.isnull().sum()>0].index[0])
print("Null prop of price column: ", s_cal.price.isnull().sum()/s_cal.shape[0])
print("Proportion of False(unit unavailable at this date ):", s_cal.available[s_cal.available =='f' ].count()/s_cal.shape[0])
#Since the proportions are the same, then, all nans are when the units aren't available, let's be more sure
print("Nulls when units are available: ", s_cal[s_cal['available']== 't']['price'].isnull().sum())


Boston Cal: 
Cols with nulls:  price
Null prop of price column:  0.5087157820748879
Proportion of False(unit unavailable at this date ): 0.5087157820748879
Nulls when units are available:  0


Seatle Cal: 
Cols with nulls:  price
Null prop of price column:  0.32938998399793334
Proportion of False(unit unavailable at this date ): 0.32938998399793334
Nulls when units are available:  0


#### Transfer Date column to datetime to ease manipulation, analysis and modeling. I create a dataframe with seperate date items from the Date column, to check the period in which the data was collected.

In [252]:
def create_dateparts(df, date_col):
    
    df['date'] = pd.to_datetime(df.date)
    print(df.date.dtype== 'datetime64[ns]')

    b_date_df = pd.DataFrame()
    b_date_df['year'] = df['date'].dt.year
    b_date_df['month'] = df['date'].dt.month
    b_date_df['day'] =df['date'].dt.day
    b_date_df['dow'] =df['date'].dt.strftime("%A")
    df = df.join(b_date_df)
    return df

In [253]:
b_cal_1 = create_dateparts(b_cal, 'date')
b_cal_1.head(3)

True


listing_id       date available price  year  month  day      dow
0    12147973 2017-09-05         f   NaN  2017      9    5  Tuesday
1    12147973 2017-09-04         f   NaN  2017      9    4   Monday
2    12147973 2017-09-03         f   NaN  2017      9    3   Sunday

In [254]:
s_cal_1 = create_dateparts(s_cal, 'date')
s_cal_1.head(3)

True


listing_id       date available   price  year  month  day        dow
0      241032 2016-01-04         t  $85.00  2016      1    4     Monday
1      241032 2016-01-05         t  $85.00  2016      1    5    Tuesday
2      241032 2016-01-06         f     NaN  2016      1    6  Wednesday

In [255]:
def get_period_df(df):
    period =pd.DataFrame(df.groupby(['year','month'], sort = True)['day'].value_counts())
    period = period.rename(columns={'day':'count'}, level=0)
    period = period.reset_index().sort_values(by=['year', 'month', 'day']).reset_index(drop = True)
    return period

In [256]:
b_period =get_period_df(b_cal_1)
b_period

year  month  day  count
0    2016      9    6   3586
1    2016      9    7   3586
2    2016      9    8   3586
3    2016      9    9   3586
4    2016      9   10   3586
..    ...    ...  ...    ...
360  2017      9    1   3586
361  2017      9    2   3586
362  2017      9    3   3586
363  2017      9    4   3586
364  2017      9    5   3586

[365 rows x 4 columns]

In [257]:
s_period =get_period_df(s_cal_1)
s_period

year  month  day  count
0    2016      1    4   3818
1    2016      1    5   3818
2    2016      1    6   3818
3    2016      1    7   3818
4    2016      1    8   3818
..    ...    ...  ...    ...
360  2016     12   29   3818
361  2016     12   30   3818
362  2016     12   31   3818
363  2017      1    1   3818
364  2017      1    2   3818

[365 rows x 4 columns]

In [258]:
print(b_period['count'].unique())
print(s_period['count'].unique())
print("Number of unique Listing IDs in Boston Calendar: ", len(b_cal_1.listing_id.unique()))
print("Number of unique Listing IDs in Seatle Calendar: ", len(s_cal_1.listing_id.unique()))

[3586]
[3818]
Number of unique Listing IDs in Boston Calendar:  3585
Number of unique Listing IDs in Seatle Calendar:  3818


#### Counts are the equivalent to the numbers of unique ids because all the ids are spanning the same time period by day.  Let's check any anomalies

In [259]:
def check_anomalies(df, col):
    list_ids_not_year_long = []

    for i in sorted(list(df[col].unique())):
        if df[df[col]== i].shape[0] != 365:
            list_ids_not_year_long.append(i)
    print("Entry Ids that don't span 1 year: " , list_ids_not_year_long)

In [260]:
#Boston
check_anomalies(b_cal_1, 'listing_id')

Entry Ids that don't span 1 year:  [12898806]


In [261]:
#Seatle
check_anomalies(s_cal_1, 'listing_id')

Entry Ids that don't span 1 year:  []


In [262]:
## check this entry in Boston Calendar
print("Span of the entries for this listing, should be 365: ", b_cal_1[b_cal_1['listing_id']== 12898806].shape[0])
## 2 years, seems like a duplicate as 730 = 365 * 2
one_or_two = pd.DataFrame(b_cal_1[b_cal_1['listing_id']==12898806].groupby(['year', 'month', 'day'])['day'].count()).day.unique()[0]
print("Should be 1: ", one_or_two)
## It indeed is :)
b_cal_1 = b_cal_1.drop_duplicates()
print("Size of anomaly listing, Should be = 365: ", b_cal_1.drop_duplicates()[b_cal_1.drop_duplicates().listing_id==12898806]['listing_id'].size)
print("After removing duplicates, Span of the entries for this listing, should be 365: ", b_cal_1[b_cal_1['listing_id']== 12898806].shape[0])
print("After removing duplicates, shape is: ", b_cal_1.shape)

Span of the entries for this listing, should be 365:  730
Should be 1:  2
Size of anomaly listing, Should be = 365:  365
After removing duplicates, Span of the entries for this listing, should be 365:  365
After removing duplicates, shape is:  (1308525, 8)


/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
### Comments:  
[Boston & Seatle Calendar]
- The datasets have information about listing dates, availability and price tracked over a year for ever listing id
- There are no data entry errors, all nulls are due to the structuring of the Data (the listings that aren't available has no price)
<br><br>
- I added 4 cols that contain dateparts that will aid further analysis and modeling
- The Boston calendar Dataset ranges through `365`days from `6th of September'16` to `5th of September'17`, No nulls with `1308525` rows and  `8` cols
- The Seatle calendar Dataset ranges through `365`days from `4th of January'16` to `2nd of January'17`, No nulls with `1393570` rows and  `8` cols
<br><br>
- Number of unique Listing IDs in Boston Calendar:  `3585`
- Number of unique Listing IDs in Seatle Calendar:  `3818`
- May need to order the table later 

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

 _______________________________________________________________________________________________________________________

## Step 1: Continue - 

> **Boston & Seatle Listings**

In [18]:
b_list.head(10)

id                            listing_url       scrape_id  \
0  12147973  https://www.airbnb.com/rooms/12147973  20160906204935   

  last_scraped                        name  \
0   2016-09-07  Sunny Bungalow in the City   

                                             summary  \
0  Cozy, sunny, family home.  Master bedroom high...   

                                               space  \
0  The house has an open and cozy feel at the sam...   

                                         description experiences_offered  \
0  Cozy, sunny, family home.  Master bedroom high...                none   

                               neighborhood_overview  ... review_scores_value  \
0  Roslindale is quiet, convenient and friendly. ...  ...                 NaN   

  requires_license license jurisdiction_names instant_bookable  \
0                f     NaN                NaN                f   

  cancellation_policy require_guest_profile_picture  \
0            moderate                             f   

  require_guest_phone_verification calculated_host_listings_count  \
0                                f                              1   

   reviews_per_month  
0                NaN  

[1 rows x 95 columns]

In [34]:
s_list.head(10)

id                          listing_url       scrape_id last_scraped  \
0  241032  https://www.airbnb.com/rooms/241032  20160104002432   2016-01-04   

                           name summary  \
0  Stylish Queen Anne Apartment     NaN   

                                               space  \
0  Make your self at home in this charming one-be...   

                                         description experiences_offered  \
0  Make your self at home in this charming one-be...                none   

  neighborhood_overview  ... review_scores_value requires_license license  \
0                   NaN  ...                10.0                f     NaN   

  jurisdiction_names instant_bookable cancellation_policy  \
0         WASHINGTON                f            moderate   

   require_guest_profile_picture require_guest_phone_verification  \
0                              f                                f   

  calculated_host_listings_count reviews_per_month  
0                              2              4.07  

[1 rows x 92 columns]

 ### Check the sizes of cols and rows 

In [94]:
print("Boston listings size :", b_list.shape)
print("Seatle listings size :", s_list.shape)
set_difference = set(b_list.columns) - set(s_list.columns)
print("Columns in Boston but not in Seatle:  ", set_difference)

Boston listings size : (3585, 95)
Seatle listings size : (3818, 92)
Columns in Boston but not in Seatle:   {'house_rules', 'access', 'interaction'}


In [93]:
b_list[['house_rules', 'access', 'interaction']].head(3)

house_rules  \
0  Clean up and treat the home the way you'd like...   
1  Pet friendly but please confirm with me if the...   
2  I encourage you to use my kitchen, cooking and...   

                                              access  \
0  You will have access to 2 bedrooms, a living r...   
1  Apt has one more bedroom (which I use) and lar...   
2  I am living in the apartment during your stay,...   

                                         interaction  
0                                                NaN  
1  If I am at home, I am likely working in my hom...  
2  ABOUT ME: I'm a laid-back, friendly, unmarried...

### Check Nulls

In [264]:
print("Number of Non-null cols in Boston listings: ",  np.sum(b_list.isnull().sum()==0), " From ", b_list.shape[1])
print("Number of Non-null cols in Seatle listings: ",  np.sum(s_list.isnull().sum()==0), " From ", s_list.shape[1])

Number of Non-null cols in Boston listings:  51  From  95
Number of Non-null cols in Seatle listings:  47  From  92


In [123]:
# For Boston Listings The null proportion per column
print(b_list.isnull().sum()[b_list.isnull().sum()>0]/b_list.shape[0])

summary                         0.039888
space                           0.294840
neighborhood_overview           0.394700
notes                           0.550907
transit                         0.359833
access                          0.415342
interaction                     0.433473
house_rules                     0.332497
thumbnail_url                   0.167085
medium_url                      0.167085
xl_picture_url                  0.167085
host_location                   0.003068
host_about                      0.365132
host_response_time              0.131381
host_response_rate              0.131381
host_acceptance_rate            0.131381
host_neighbourhood              0.094561
neighbourhood                   0.151464
neighbourhood_group_cleansed    1.000000
city                            0.000558
zipcode                         0.010600
market                          0.003905
property_type                   0.000837
bathrooms                       0.003905
bedrooms        

In [293]:
# For Seatle Listings The null proportion per column
#s_list.isnull().sum()[s_list.isnull().sum()>0]/s_list.shape[0]*100

# Examine columns closely to prep them correctly!

In [ ]:
#To be removed

# listing_url, scrape_id, last_scraped, name, experiences_offered, 
thumbnail_url,
medium_url,
picture_url,
xl_picture_url
#host_id, host_url, host_thumbnail_url, host_picture_url, host_total_listings_count, neighbourhood, neighbourhood_group_cleansed,
# state, country_code, country, latitude, longitude , has_availability, calendar_last_scraped, requires_license,
#license, jurisdiction_names

#To check :

#host_since      ++, 
#host_location     +++, 
#host_about           + , 
#host_response_time ++++,  
#host_response_rate   +++++ "Transform to numeric", 
#host_acceptance_rate  +++++ "Transform to numeric", 
#host_is_superhost     +++++ "Transform to Boolean???", 
#host_neighbourhood   +++++    "What is neighborhood/cleansed then????"
#host_listings_count   +++++
#host_verifications    +++       "List of strings may need ot be broken down "
#host_has_profile_pic ++    "Transform to Boolean???"
#host_identity_verified      "Transform to Boolean???"
#neighbourhood_cleansed ++++ "What is neighborhood then"
#city   ++++                   "39 unique values"
#zipcode ++++                  "44 unique values, Not as many as expected"
#market  +++                   "I don't understand what market is? Boston', 'San Francisco', nan, 'Other (Domestic)'Other (International)'"
#smart_location  ++++           '39 unique values '
#is_location_exact ++++         "Transform to Boolean???"
#property_type    ++++           "propert type apartment, house, etc. "
#room_type         ++++           "'Entire home/apt', 'Private room', 'Shared room'"
#accommodates      ++++            "INteger"
#bathrooms          ++++            "Float"
#bedrooms         ++++               "Float"
#beds                ++++             "Float"
#bed_type      ++++                  "Real Bed', 'Futon', 'Pull-out Sofa', 'Airbed', 'Couch'
#amenities     +++                    "Lomg text needs to be broken down"
#square_feet +++                       "more thna 90% null___ sad, drop probably"
#price      +++++                       Öbject, transform to int/float"  
#weekly_price +++++   *                   75% Null, Öbject, transform to int/float" 
#monthly_price +++++  *                  75% Null,  Öbject, transform to int/float" 
#security_deposit +++                    65% Null,
#cleaning_fee  ++++                      30% null 
#guests_included +++++                    "INteger"
#extra_people   ++++                      object, transform to int/float"
#minimum_nights +++++                     ÏNteger"
#maximum_nights   ++   *                   "needs work on it Maybe drop maybe divide into ranges"
#calendar_updated ++                        "Since when was the calendar updated"
#availability_30
#availability_60
#availability_90
#availability_365                        "INtegers"
#number_of_reviews                         "INteger"
#first_review                               "date, WHen was first review"
#last_review                                 "date, WHen was last review"
#review_scores_rating                         " 20% nans, floats, INvestigate what is this ?? "
#review_scores_accuracy                       " 20% nans, floats"
#review_scores_cleanliness                        "20% nans, float"
#review_scores_checkin                        " 20% nans, floats"
#review_scores_communication                  " 20% nans, floats"
#review_scores_location                       " 20% nans, floats"
#review_scores_value                            " 20% nans, floats"
#instant_bookable                              "Transform to Boolean???"
#cancellation_policy                           'moderate', 'flexible', 'strict', 'super_strict_30'
#require_guest_profile_picture                  "Transform to Boolean???"
#require_guest_phone_verification                "Transform to Boolean???"
#calculated_host_listings_count                   "int"
#reviews_per_month                                 "20% nans, FLoat, "

#host_name - , 
#summary -             >> short text description of the listing      >>  length, text content ??
#description *            >> long text description of the listing       >>                         //
#neighborhood_overview * >> short text description of the neighborhood >>                        //
#notes *                >> short text with some notes                 >>                         //
#transit -             >> short text describing the transportation   >>                       //
#access **               >> only in boston,  short text describing what the guest can access >> //
#interaction **          >> only in boston, short text about guest and owner interaction >>      //
#house_rules  **        >> only in boston, short text about house rules >>                     //
#street -


In [266]:
b_list.columns

Index(['id', 'listing_url', 'scrape_id', 'last_scraped', 'name', 'summary',
       'space', 'description', 'experiences_offered', 'neighborhood_overview',
       'notes', 'transit', 'access', 'interaction', 'house_rules',
       'thumbnail_url', 'medium_url', 'picture_url', 'xl_picture_url',
       'host_id', 'host_url', 'host_name', 'host_since', 'host_location',
       'host_about', 'host_response_time', 'host_response_rate',
       'host_acceptance_rate', 'host_is_superhost', 'host_thumbnail_url',
       'host_picture_url', 'host_neighbourhood', 'host_listings_count',
       'host_total_listings_count', 'host_verifications',
       'host_has_profile_pic', 'host_identity_verified', 'street',
       'neighbourhood', 'neighbourhood_cleansed',
       'neighbourhood_group_cleansed', 'city', 'state', 'zipcode', 'market',
       'smart_location', 'country_code', 'country', 'latitude', 'longitude',
       'is_location_exact', 'property_type', 'room_type', 'accommodates',
       'bathrooms',

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
### Comments:  
[Boston & Seatle Listings]
- Boston listings size : `3585`, `95`
- Seatle listings size : `3818`, `92`
- Boston Listings has 3 cols that arent in Seatle, `house_rules`, `access`, `interaction`, types: `Object/String`
- Number of Non-null cols in Boston listings:  `51`, around half
- Number of Non-null cols in Seatle listings:  `47`, around half
